File: src/main/java/com/rwbymod/client/RenderRegistry.java
package com.rwbymod.client;

import com.rwbymod.block.BlockRegistry;
import com.rwbymod.item.ItemRegistry;
import com.rwbymod.misc.EnumDustType;

import net.minecraft.block.Block;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.client.renderer.color.IBlockColor;
import net.minecraft.client.renderer.color.IItemColor;
import net.minecraft.item.Item;

public class RenderRegistry {
	
	public static void init() {
		registerSimpleMetaItemRenderer(ItemRegistry.dust, EnumDustType.values().length);
		registerItemColorHandler(new ColorizerDust(), ItemRegistry.dust);

		registerSimpleMetaBlockRenderer(BlockRegistry.dustOre, EnumDustType.primary.size());
		registerBlockColorHandler(new ColorizerDust(1), BlockRegistry.dustOre);
		registerItemColorHandler(new ColorizerDust(1), Item.getItemFromBlock(BlockRegistry.dustOre));
	}
	
	// ----- UTILITY FUNCTIONS ----- //
	
	private static void registerItemRenderer(Item item, int meta) {
		Minecraft.getMinecraft().getRenderItem().getItemModelMesher().register(item, meta, new ModelResourceLocation(item.getRegistryName(), "inventory"));
	}
	private static void registerItemRenderer(Item item) { registerItemRenderer(item, 0); }
	
	private static void registerSimpleMetaItemRenderer(Item item, int metaCount) {
		for(int i = 0; i < metaCount; i++) registerItemRenderer(item, i);
	}
	
	private static void registerItemColorHandler(IItemColor color, Item item) {
		Minecraft.getMinecraft().getItemColors().registerItemColorHandler(color, item);
	}
	
	private static void registerBlockRenderer(Block block, int meta) {
		Item item = Item.getItemFromBlock(block);
		Minecraft.getMinecraft().getRenderItem().getItemModelMesher().register(item, meta, new ModelResourceLocation(item.getRegistryName(), "inventory"));
	}
	private static void registerBlockRenderer(Block block) { registerBlockRenderer(block, 0); }
	
	private static void registerSimpleMetaBlockRenderer(Block block, int metaCount) {
		for(int i = 0; i < metaCount; i++) registerBlockRenderer(block, i);
	}
	
	private static void registerBlockColorHandler(IBlockColor color, Block block) {
		Minecraft.getMinecraft().getBlockColors().registerBlockColorHandler(color, block);
	}
}

File: src/main/scala/com/rwbymod/registry/RenderRegistry.scala
package com.rwbymod.registry

import net.minecraft.block.Block
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.client.renderer.color.IBlockColor
import net.minecraft.client.renderer.color.IItemColor
import net.minecraft.item.Item
import com.rwbymod.dust.EnumDustType
import com.rwbymod.dust.ColorizerDust

object RenderRegistry {

  def init {
    register(ItemRegistry.dust, 0 to EnumDustType.all.size)
    register(ItemRegistry.dustCrystal, 0 to EnumDustType.all.size)
    register(ItemRegistry.scroll)
    register(BlockRegistry.dustOre, 0 to EnumDustType.primary.size)

    registerColorizer(ItemRegistry.dust, new ColorizerDust())
    registerColorizer(ItemRegistry.dustCrystal, new ColorizerDust())
    registerColorizer(BlockRegistry.dustOre, new ColorizerDust(1))
  }

  // --- UTILITY METHODS --- //

  def register(in:Item):Unit = register(in, 0)
  def register(in:Block):Unit = register(in, 0)
  def register(in:Item, meta:Int):Unit =
    Minecraft.getMinecraft.getRenderItem.getItemModelMesher.register(in, meta, new ModelResourceLocation(in.getRegistryName(), "inventory"))
  def register(in:Block, meta:Int):Unit =
    Minecraft.getMinecraft.getRenderItem.getItemModelMesher.register(Item.getItemFromBlock(in), meta, new ModelResourceLocation(in.getRegistryName(), "inventory"))
  def register(in:Item, meta:Range):Unit = meta.foreach { m => register(in, m) }
  def register(in:Block, meta:Range):Unit = meta.foreach { m => register(in, m) }

  def registerColorizer(item:Item, color:IItemColor) = Minecraft.getMinecraft.getItemColors.registerItemColorHandler(color, item)
  def registerColorizer(block:Block, color:(IBlockColor with IItemColor)):Unit = registerColorizer(block, color, color)
  def registerColorizer(block:Block, blockColor:IBlockColor, itemColor:IItemColor):Unit = {
    Minecraft.getMinecraft.getBlockColors.registerBlockColorHandler(blockColor, block)
    registerColorizer(Item.getItemFromBlock(block), itemColor)
  }

}